src/components/SearchBar.tsx

```diff
@@ -0,0 +1,91 @@
+import * as React from 'react';
+import * as _ from 'underscore';
+
+export interface SearchBarProps {
+    onTextChange: Function;
+    onStructuredModeChanged: Function;
+    onDetailedModeChanged: Function;
+    onCategoriesChange: Function;
+    categories: string[];
+}
+
+export interface SearchBarState { expanded: boolean; selectedCategories: string[] }
+
+export class SearchBar extends React.Component<SearchBarProps, SearchBarState> {
+
+    constructor(props: SearchBarProps) {
+        super(props);
+        this.state = { expanded: false, selectedCategories: [] };
+    }
+
+    onTextChange(event: any) {
+        this.props.onTextChange(event);
+    }
+
+    onOptionsButtonClick() {
+        this.setState({ expanded: !this.state.expanded });
+    }
+
+    onStructuredModeChange(event: any) {
+        this.props.onStructuredModeChanged();
+    }
+
+    onDetailedModeChange(event: any) {
+        this.props.onDetailedModeChanged();
+    }
+
+    onCategoriesChange(event: any) {
+        let categories = this.state.selectedCategories;
+        if (_.contains(this.state.selectedCategories, event.target.name)) {
+            categories = _.without(categories, event.target.name);
+        }
+        else {
+            categories.push(event.target.name);
+        }
+        this.setState({ selectedCategories: categories})
+        this.props.onCategoriesChange(categories);
+    }
+
+    render() {
+        let options = <div></div>;
+        let searchOptionsBtn = <button id="SearchOptionsBtn" className="ArrowBg" onClick={this.onOptionsButtonClick.bind(this)}></button>;
+        let thisObj = this;
+        var checkboxes = [];
```

Comment:
It's better to decalre the type of a variable as much as possible, so it should be `JSX.Element[]`